### DISCUSSION TITLE
Nested `TouchArea`: catch the same event at different levels.

### DISCUSSION THREAD

**Discussion by ghost** [slint-ui/slint#3255] [Q&A] [answered]
Hi! 🙂 I'm implementing a screensaver for my `C++`/`Slint` app, which is running in kiosk mode. I would like to trigger the screensaver after a given idle period, aka after some time without any incoming touch event. Basically I would like to do something like that:

``` slint
TouchArea {
    clicked => { debug("also catched"); }
    TouchArea {
        clicked => { debug("catched"); }
    }
}
```

When I try this, only the inner-most `TouchArea` gets notified. How can I make it working? Alternatively, is there a way to listen to touch events globally from `C++`?

Thanks! 🙂

**Comment by tronical** [comment]
I think @FloVanGH was trying to solve a similar issue for the energy monitor demo. How did you do that again?

My naive and untested approach would be to place a touch area at the bottom:

```
export component App {
    // ... all your UI elements here

    // This comes last and will span the entire window
    if screen-saver-enabled: TouchArea {
        clicked => { root.screen-saver-enabled = false; }
    }
}
```

I made a little [example on SlintPad](https://slintpad.com/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++property+%3Cbool%3E+screen-saver%3B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++%7D%0A++++++++Image+%7B%0A++++++++++++source%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++height%3A+100px%3B%0A++++++++%7D%0A++++++++HorizontalLayout+%7B+alignment%3A+center%3B+Button+%7B+text%3A+%22Enable+screen+saver%22%3B+clicked+%3D%3E+%7B+screen-saver+%3D+true%3B+%7D+%7D+%7D%0A++++%7D%0A++++if+screen-saver%3A+TouchArea+%7B%0A++++++++clicked+%3D%3E+%7B+screen-saver+%3D+false%3B+%7D%0A++++++++Text+%7B%0A++++++++++++text%3A+%22screen+saver+on%22%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A).

**Reply by ghost** [reply]
Thanks for your answer. 🙂 I played around with this solution a bit, but that doesn't work for my use case: I would like to get notified each time the root window gets an incoming touch event (or any user input event), without preventing that event from reaching its destination. I can workaround this by calling the same callback in every `TouchArea` of my application, but it feels a bit heavy. Putting `TouchArea` at the bottom of the root component catches everything, but it cancel the event propagation. Nesting `TouchArea` doesn't prevent propagation, but the top-level one only catches events that are not caught by lower-level ones.

**Reply by tronical** [reply]
Ahh right, the proposed suggestion intercepts the event, but you want it to go still be processed, right?

Perhaps it would suffice to have an `accept` field in the `pointer-event`'s `event` argument, default it to `true`, but allow you to set it to false - so that you see the event but you can say "I don't accept it" and thus it'll be propagated to what's underneath the area.

Olivier, you're the TouchArea expert :-). Does this make sense?

**Reply by ogoffart** [reply]
We're missing such API, yes.
Some option on how we could implement this feature:
 1. We could change the return type of the pointer-event callback,  and have a feature that some enum can be constructed from void in return position.
 2. we could add a property to the touch area `transparent`  or `propagate` 
 3. We could use a different element that have different behavior like a gesture recognizer or something like that.

**Reply by tronical** [reply]
I think (1) is the cleanest if it's combined with a warning that this conversion will be removed in say "2.0" and that one must specify accept or ignore (like for the focus scope).

**Reply by ogoffart** [reply]
So if you want to have a Transparent TouchArea you'd do

``` 
// use ta.pressed and other properties in the rest of the code
ta := TouchArea {
    clicked => { debug("clicked") }
    
    //but in order to make it transparent, add this callback
    pointer-event => { return reject }
}
```

But the problem is that EventResult::reject is already currently the default value of that enum.

> combined with a warning

We will get lots and lots of warning.  Also i find it less ergonomic in the typical case to always have to return something

**Comment by ghost** [comment, marked as answer]
Thanks for your interest in my use case! In the meantime I think I will simply redefine my own `TouchArea` with a custom hook in `clicked`, and use this `MyTouchArea` everywhere. 🙂
